### PR TITLE
Fixes http header retrieval issue in `Metadata` reported in #26551

### PR DESCRIPTION
**Pull request by StingyJack**
thus complying with rfc2616 section 4.2 
also rfc7230 section 3.2 
and rfc7540 section 8.1.2

#26551 

@yashykt

### CODE REVIEW

**Comment by StingyJack** [comment]
The csharp solution does not compile after cloning the repo and following the readme `('cmake' is not recognized as an internal or external command,
operable program or batch file.)`.  I dont see any unit tests that would be affected that need to be fixed as well.  This is just a correction to a string comparison.

I dont know what this status check means or what it wants. 
image

**Comment by StingyJack** [comment]
@yashykt any insight into what I need to do to satisfy the mergeable status check?

**Comment by yashykt** [comment]
@StingyJack Sorry for the delay!

**Comment by StingyJack** [comment]
np @yashykt, I have a workaround.   However I'm not sure what to do about the failing status checks, I wont have a mac available to do any testing with, and the sanity check failed because it couldnt find a python module (`ModuleNotFoundError: No module named '_bz2'`). I'm pretty sure I didnt do anything to cause that.

**Comment by yashykt** [comment]
@StingyJack Ignore the failures. They are unrelated to your change.

**Comment by jtattermusch** [comment]
Thanks for proactively trying to fix a potential source of confusion, but I'm not sure if this is a correct fix. The problem is that this modifies behavior of an already-existing public method and there might be user code somewhere that relies on the existing behavior. Whether the old behavior is wrong and the new behavior is correct is something that's subjective. To avoid disruption, a more conservative choice might be to just improve the documentation and clearly describe the current behavior.

Btw, is there any good reason why you're working with mixed-case header names in your app. Since gRPC converts all the header names to lowercase anyway, you can simply pass only lower case values as metadata entry names and you won't have any problems.

**Comment by StingyJack** [comment]
> is there any good reason why you're working with mixed-case header names in your app.

``` c#
public const string OneReasonIs  = "BecauseHumansReadSourceCode"; // snake_case_not_an_option
```

... and the actual code that prompted this. `UserExecutionContext` is a type, but `nameof()` works for pretty much anything
``` c#
//the .ToLowerInvariant() is due to a defect in the Grpc Metadata class. I've submitted https://github.com/grpc/grpc/pull/26552 to fix it.
var uecRaw = serverCallContext.RequestHeaders.Get($"{nameof(UserExecutionContext)}{Metadata.BinaryHeaderSuffix}".ToLowerInvariant());
if (uecRaw == null)
{
     throw new InvalidOperationException($"Unable to get the {nameof(UserExecutionContext)} from the request headers");
}
```

> The problem is that this modifies behavior of an already-existing public method and there might be user code somewhere that relies on the existing behavior

Grammar rules aside, is there a difference in meaning between "The dog barks" and "The Dog Barks" ?  No, because humans dont assign different meaning to a word when the case differs. If they did, the http standard wouldn't allow the headers to be forced to lowercase on transmit. 

I dont see how this would break an existing implementation. Can you give a reasonably possible, concrete, non-theoretical example please? 

[The relevant spec from IETF](https://www.rfc-editor.org/rfc/rfc7540.html#section-8.1.2).
```
8.1.2.  HTTP Header Fields

   HTTP header fields carry information as a series of key-value pairs.
   For a listing of registered HTTP headers, see the "Message Header
   Field" registry maintained at <https://www.iana.org/assignments/
   message-headers>.

   Just as in HTTP/1.x, header field names are strings of ASCII
   characters that are compared in a case-insensitive fashion.  However,
   header field names MUST be converted to lowercase prior to their
   encoding in HTTP/2.  A request or response containing uppercase
   header field names MUST be treated as malformed (Section 8.1.2.6).
```

**Comment by StingyJack** [comment]
>  Whether the old behavior is wrong and the new behavior is correct is something that's subjective. 

Also @jtattermusch  this is not a matter of subjectivity or grey area. The spec is very clear on this point. 

Before the proposed change, if I added a header with a key of "SomeHttpHeaderKey" when sending a message, the receiver could not retrieve the header value using the same "SomeHttpHeaderKey" and would have to use "somehttpheaderkey". Whereas after these proposed changes, the same header value can be retrieved by the receiver using any casing; "SomeHttpHeaderKey", "somehttpheaderkey", "SOMEhttpHEADERkey", etc.   How exactly does the proposed change break someone's code?

**Comment by StingyJack** [comment]
@yashykt @markdroth  - can this be reassigned to someone else to review? This fixes something that is very clearly not compliant with the HTTP spec and makes it compliant. 

There has been no problem identified with these changes, none of the tests report failure after this change, and neither the reviewer or I can come up with a concrete legitimate scenario where this change could break anything. I'm not sure what the current reviewer's hesitation is, but at this point its clear they do not intend to act on this one way or another.

**Comment by jtattermusch** [comment]
> @yashykt @markdroth - can this be reassigned to someone else to review? This fixes something that is very clearly not compliant with the HTTP spec and makes it compliant.
> 
> There has been no problem identified with these changes, none of the tests report failure after this change, and neither the reviewer or I can come up with a concrete legitimate scenario where this change could break anything. I'm not sure what the current reviewer's hesitation is, but at this point its clear they do not intend to act on this one way or another.

@StingyJack  nice try bypassing me ;-)

Let's not invoke HTTP/2 spec compliance, as the problem you're seeing has nothing to do with it.
Note that the fact that gRPC's "metadata" are transferred on the wire using HTTP/2 headers/trailers is an implementation details. The "metadata" (as defined in   In https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md) is a gRPC specific concept that is independent of headers/trailers HTTP/2 spec and their similarity is a "concidence".

The gRPC spec (https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md) clearly states that the metadata keys are only allowed to use lowercase characters (see grammar: `Header-Name → 1*( %x30-39 / %x61-7A / "_" / "-" / ".") ; 0-9 a-z _ - .`).  So whenever a gRPC implementation (grpc c# in our case) allows you to use upper-case characters in your code, it can be considered a convenience feature / syntactic sugar, but as said is in no way related to HTTP/2 spec.

The way I look at the problem reported in https://github.com/grpc/grpc/issues/26551 is that it's a gRPC C# specific API oddity and I agree it can result in some confusion in some situations, but the situations in which it exihibits itself aren't comment, are completely preventable and it doesn't influence the correctness of gRPC's behavior in any way (as long as you use lowercase metadata keys, as you should according to the grpc spec). 

I'm trying to figure out what's the best way of addressing the API oddity you described and since IMHO the answer isn't as clear cut as you say, it's taking time (also because this is not really a high-priority issue as you are making it seem, but rather a cleanup item).

I hope we're on the same page now.

File: src/csharp/Grpc.Core.Api/Metadata.cs [hunk 1]
```diff
@@ -83,7 +83,7 @@ public Entry Get(string key)
         {
             for (int i = entries.Count - 1; i >= 0; i--)
             {
-                if (entries[i].Key == key)
+                if (string.Equals(entries[i].Key, key, StringComparison.OrdinalIgnoreCase))
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core.Api/Metadata.cs:86]: since NormalizeKey()  uses `key.ToLowerInvariant();`, we'd need to use StringComparison.InvariantCultureIgnoreCase here.

**Comment by jtattermusch** [comment]
Created https://github.com/grpc/grpc/pull/27383

**Comment by StingyJack** [comment]
https://github.com/grpc/grpc/commit/d64f75d0294b0f7d383033e06e9fcbecf3d6a926

@jtattermusch - I asked for a different reviewer because you had decided to ignore my PR because you were afraid it would break something. I took the time and effort to explain how it would not in several posts above and you continued to ignore a fix for where these gRPC tools was not following the HTTP spec.  After a few weeks I asked for this to be assigned to a different reviewer because you were continuing to ignore it. 

You have deleted the comment requesting the reassignment, It was not abusive, or hate speech, or criminal, or a violation of any Code of Conduct provision. It wasnt even rude, and it could very easily have been by just reflecting what has been cast toward me.

And then rather than accept these changes, you made your own changes to do (almost, but not quite) the same thing. Aren't you afraid of your changes breaking some existing usages as you were of mine? Mine arent going to throw a NullReferenceException if a key is null. Mine are also considerably simpler, clearer, and more conventional, so there will be less chance of any later maintenance gaffes.  I think that you found mine unsatisfactory for no other reason except personal pride and that, along with deleting my comment here, is completely unprofessional and unethical and _is_ a violation of the CoC.  

You didn't even have the decency to leave a note or explanation, just a merge conflict on this PR and a missing comment. 

CC: @yashykt @markdroth and taking a screenshot in case JT deletes my comment again.  And since working in this repo is probably part of your day job duties, once I can find where to report this conduct to google, it _will_ be reported.

File: src/csharp/Grpc.Core.Api/Metadata.cs (hunk 1 above)
Comment (StingyJack) [reply src/csharp/Grpc.Core.Api/Metadata.cs:86]: String conversion isnt the same as string comparison. If this were using different encodings I would agree but its not and these do not have to "match" in that way. 

Also there isnt a .ToLowerOrdinal() function so .ToLowerInvariant() is the best available choice.  OrdinalIgnoreCase and InvariantCultureIgnoreCase both eschew culture specifics, but Ordinal is preferred for [several reasons](https://docs.microsoft.com/en-us/dotnet/standard/base-types/best-practices-strings). 

And your commit also uses OrdinalIgnoreCase (albeit in an odd and hard to follow manner), so why not resolve or remove this comment?

**Comment by jtattermusch** [comment]
Superseded by #27383 which:
 - introduces a single place where keys are compared for equality (clearer if in the future there are more places where we need to compare key or if we need to change the way the keys are compared).
 - explains in a comment why StringComparison.OrdinalIgnoreCase is fine to use (since lowercasing is done with .ToLowerInvariant()).
 - adds tests.
